### PR TITLE
Immediately try again to stop countdown on screen change

### PR DESCRIPTION
**Pull request by StefanBruens**
The current way of resubmitting the stop keycombo (Alt-s) when it was
missed is problematic for two reasons:
- A retry only happens after the 4 second timeout, not after 2.5 seconds,
  as the wait_still_screen keeps watching the screen up to the timeout
- The timeout is scaled by the TIMEOUT_SCALE, i.e. with a scale of 3 or
  more no retry happens.

Try to mimic the behaviour of a human operator - as soon as a screen
change happens, resubmit the key combo, if no change happens for 3 seconds
assume the countdown has stopped.

Signed-off-by: Stefan Brüns <stefan.bruens@rwth-aachen.de>

wait_screen_change: https://github.com/os-autoinst/os-autoinst/blob/master/testapi.pm#L444
wait_still_screen https://github.com/os-autoinst/os-autoinst/blob/master/testapi.pm#507

### CODE REVIEW

File: tests/installation/install_and_reboot.pm [hunk 1]
```diff
@@ -36,6 +36,13 @@ sub handle_livecd_screenlock {
     assert_screen('yast-still-running', 120);
 }
 
+sub wait_countdown_stop {
+    my $stilltime = shift;
+    return wait_screen_change ( sub {
+        send_key 'alt-s';
+    }, $stilltime);
```
Comment (okurz) [inline tests/installation/install_and_reboot.pm]: I guess it should just be `return wait_screen_change { send_key 'alt-s' }, $stilltime`

**Comment by StefanBruens** [comment]
I have run it twice locally, but times succeeded:

```
16:46:08.1538 679 no match: 4900.5s
16:46:09.0163 679 MATCH(rebootnow-20150127:0.23)
16:46:09.7814 679 MATCH(rebootnow-20151116:0.94)
16:46:10.5452 679 MATCH(rebootnow-20161215:1.00)
16:46:12.1666 679 MATCH(yast2_wrong_digest:0.00)
16:46:13.0492 679 MATCH(yast2_wrong_digest_textmode:0.00)
16:46:21.4015 675 >>> testapi::_handle_found_needle: found rebootnow-20161215, similarity 1.00 @ 271/261
16:46:21.4860 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:43 called testapi::wait_screen_change
16:46:21.5505 675 <<< testapi::wait_screen_change(timeout=3)
16:46:21.6132 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:42 called testapi::send_key
16:46:21.6151 675 <<< testapi::send_key(key='alt-s')
16:46:22.4887 675 waiting for screen change: 0 1000000
16:46:23.0033 675 waiting for screen change: 1 1000000
16:46:23.5166 675 waiting for screen change: 1 1000000
16:46:24.0298 675 waiting for screen change: 2 1000000
16:46:24.5435 675 waiting for screen change: 2 1000000
16:46:26.3526 675 >>> testapi::wait_screen_change: timed out
16:46:26.3622 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:133 called testapi::select_console
16:46:26.3631 675 <<< testapi::select_console(testapi_console='install-shell')
16:46:26.8255 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:133 called testapi::select_console
16:46:26.8269 675 <<< testapi::assert_screen(mustmatch='install-shell', timeout=30, no_wait=1)
16:46:41.4525 679 load of /var/lib/openqa/share/tests/opensuse/products/opensuse/needles/install-shell-no_cursor-20170321.png took 0.16 seconds
16:46:41.6491 679 MATCH(install-shell-no_cursor-20170321:1.00)
```

And an earlier run (different timeout, different message, but otherwise the same):
```
01:50:32.4966 7101 MATCH(rebootnow-20150127:0.23)
01:50:32.5557 7101 MATCH(rebootnow-20151116:0.94)
01:50:32.6150 7101 MATCH(rebootnow-20161215:1.00)
01:50:32.6866 7101 MATCH(yast2_wrong_digest:0.00)
01:50:32.7129 7101 MATCH(yast2_wrong_digest_textmode:0.00)
01:50:40.5572 7080 >>> testapi::_handle_found_needle: found rebootnow-20161215, similarity 1.00 @ 271/261
01:50:40.6411 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:43 called testapi::wait_screen_change
01:50:40.7101 7080 <<< testapi::wait_screen_change(timeout=5)
01:50:40.7727 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:42 called testapi::send_key
01:50:40.7755 7080 <<< testapi::send_key(key='alt-s')
01:50:42.6701 7080 waiting for screen change: 0 51.0677058570692
01:50:43.1887 7080 waiting for screen change: 1 20.2718754133288
01:50:43.1943 7080 >>> testapi::wait_screen_change: screen change seen at 1
01:50:43.2006 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:145 called testapi::record_info
01:50:43.2022 7080 <<< testapi::record_info(title='workaround', output='While trying to stop countdown no still screen could be detected, retrying up to 9 times more', result='ok')
01:50:43.2545 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:43 called testapi::wait_screen_change
01:50:43.2567 7080 <<< testapi::wait_screen_change(timeout=5)
01:50:43.2732 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:42 called testapi::send_key
01:50:43.2741 7080 <<< testapi::send_key(key='alt-s')
01:50:43.6307 7080 waiting for screen change: 0 1000000
01:50:44.1594 7080 waiting for screen change: 1 1000000
01:50:44.6741 7080 waiting for screen change: 1 6.16152786575771
01:50:44.6747 7080 >>> testapi::wait_screen_change: screen change seen at 1
01:50:44.6753 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:145 called testapi::record_info
01:50:44.6764 7080 <<< testapi::record_info(title='workaround', output='While trying to stop countdown no still screen could be detected, retrying up to 8 times more', result='ok')
01:50:44.6781 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:43 called testapi::wait_screen_change
01:50:44.6788 7080 <<< testapi::wait_screen_change(timeout=5)
01:50:44.6907 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:42 called testapi::send_key
01:50:44.6915 7080 <<< testapi::send_key(key='alt-s')
01:50:44.9170 7080 waiting for screen change: 0 1000000
01:50:45.4300 7080 waiting for screen change: 1 1000000
01:50:45.9435 7080 waiting for screen change: 1 1000000
01:50:46.4577 7080 waiting for screen change: 2 1000000
01:50:46.9974 7080 waiting for screen change: 2 1000000
01:50:47.5124 7080 waiting for screen change: 3 1000000
01:50:48.0253 7080 waiting for screen change: 4 1000000
01:50:48.5389 7080 waiting for screen change: 4 1000000
01:50:58.2786 7080 >>> testapi::wait_screen_change: timed out
01:50:58.2921 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:147 called testapi::select_console
01:50:58.2930 7080 <<< testapi::select_console(testapi_console='install-shell')
01:50:58.6696 Debug: /var/lib/openqa/share/tests/opensuse/tests/installation/install_and_reboot.pm:147 called testapi::select_console
01:50:58.6710 7080 <<< testapi::assert_screen(mustmatch='install-shell', no_wait=1, timeout=30)
01:51:04.7398 7101 load of /var/lib/openqa/share/tests/opensuse/products/opensuse/needles/install-shell-no_cursor-20170321.png took 0.43 seconds
01:51:04.8290 7101 MATCH(install-shell-no_cursor-20170321:1.00)
```

File: tests/installation/install_and_reboot.pm (hunk 1 above)
Comment (StefanBruens) [inline tests/installation/install_and_reboot.pm]: AFAIK the 'sub' is strictly required, otherwise perl does not recognize it as an anonymous function.

**Comment by okurz** [comment]
your approach to stop the reboot countdown sounds very promising. Let other reviewers after the weekend have there chance. If there is no response until e.g. Monday evening I will merge it and we'll see if it is more robust than what we currently have.

**Comment by sysrich** [comment]
Belatedly, but +100, thank you SO MUCH for this PR - if it works as I think it will, you have just fixed a huge problem that was affecting SLE QA as much as openSUSE .

Thank you.

(and thanks to @asmorodskyi for beating me to merging before my morning coffee ;))

**Comment by okurz** [comment]
Just looking at the most recent finished results I can see it works flawless in e.g.
* https://openqa.opensuse.org/tests/407212
* https://openqa.opensuse.org/tests/407213
* https://openqa.opensuse.org/tests/407204

and these two with a single workaround window each:
* https://openqa.opensuse.org/tests/407215#step/install_and_reboot/2
* https://openqa.opensuse.org/tests/407202

so maybe a bit too eager and not patient enough but also "good enough" for now in case you don't want to revisit.

**Comment by StefanBruens** [comment]
I think it sees the "button press". Maybe just do the send_key outside the wait_screen_change callback? But even then, we might see the button release ...

**Comment by okurz** [comment]
Yes, I figured the same. It's really tricky and the only idea I can come up with is to extend `wait_screen_change` and `wait_still_screen` with a feature to "do not/only look at certain parts of the screen", e.g. a needle for `wait_screen_change`, `wait_still_screen` … in case you are looking for a summer project ;-)

**Comment by kravciak** [comment]
It's generating many soft-failures, my guess is:
 - pressing alt-s itself triggers screen_change
 - wait_screen_change - default "similarity_level 50" is sometimes not enough to match timer change

Examples:
 - http://dhcp91.suse.cz/tests/6530#step/install_and_reboot/7
 - https://openqa.suse.de/tests/1037799#step/install_and_reboot/4

**Comment by okurz** [comment]
@kravciak that is not a soft failure but just an informative window. The module is passed. If you think that is still too noisy you could just replace the record_info by a diag line or go even further because I think your first assumption is true: "alt-s" presses the button and the release of the button is seen as a screen change so maybe the better way would be to just get rid of the record_info for good.

**Comment by kravciak** [comment]
@okurz - please check https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/3216

**Pull request merged**